src/stash.c

```diff
@@ -159,7 +164,37 @@ struct stash_update_rules {
 	bool include_ignored;
 };
 
+/* similar to git_index_add_bypath but able to operate on any
+ * index without making assumptions about the repository's index
+ */
+static int stash_to_index(
+	git_repository *repo,
+	git_index *index,
+	const char *path)
+{
+	git_index *repo_index;
+	git_index_entry entry = {{0}};
+	struct stat st;
+	int error;
+
+	if (!git_repository_is_bare(repo) &&
+		(error = git_repository_index__weakptr(&repo_index, repo)) < 0)
+		return error;
```

Comment:
Does this even have to be fatal? You're only using this to construct whether we want to trust the filemode or not, so in case we cannot use the repo index we can just ignore it.

```
int trust_filemode = 1;

if (!git_repository_is_bare(repo) && !git_repository_index__weakptr(&repo_index, repo))
     trust_filemode = !repo_index->distrust_filemode;
...
git_index_entry__init_from_stat(&entry, &st, trust_filemode);

Reply:
It doesn't really _need_ to be fatal, per se.  But I expect anybody calling stash to need an index for their repository, so I think failing should not be harmful, and it's easier to reason about what happens if the index cannot be opened for some reason.